sublime/pythonLiveCoding.py

```diff
@@ -0,0 +1,306 @@
+            'cols': cols, 
+            'rows': rows, 
+            'cells': cells
+        })
+
+        # Create a new file in the new group, then switch active group back.
+        self.window.focus_group(len(cells) - 1)
+        output_view = self.window.new_file()
+        self.window.focus_group(active_group)
+        logging.getLogger().info('Created output view id: {}'.format(output_view.id()))
+
+        return output_view
+
+
+class ResetCommand(BaseWindowCommand):
+
+    def run(self):
+        
+        # Close any output views and remove their tag.
+        for view in self.window.views():
+            if not view.settings().has(OUTPUT_VIEW_ID):
+                continue
+            output_view_id = view.settings().get(OUTPUT_VIEW_ID)
+            output_view = find_view(output_view_id)
+            if output_view is not None:
+               logging.getLogger().info('Closing view: {}'.format(output_view.name()))
+               output_view.close()
+            view.settings().erase(OUTPUT_VIEW_ID)
+
+        # Set the layout back to a single group.
+        rows, cols, cells = self.get_layout()
+        self.fixed_set_layout(self.window, {
+            'cols': [0, 1], 
+            'rows': [0, 1], 
+            'cells': [[0, 0, 1, 1]]
+        })
+
+
+class BaseStartCommand(BaseWindowCommand):
+
+    TRACER_ARGS = []
+
+    def run(self):
+        active_group = self.window.active_group()
+        input_view = self.window.active_view_in_group(active_group)
+        if input_view.settings().has(OUTPUT_VIEW_ID):
+            msg = 'Already a live coding session for the current view.'
+            sublime.message_dialog(msg)
+            return
+
+        output_view = self.create_pane()
+        output_view.set_scratch(True)
+        output_view.settings().set('word_wrap', False)
+        output_view.set_name('live coding output')
+
+        input_view.settings().set('word_wrap', False)
+        input_view.settings().set(OUTPUT_VIEW_ID, output_view.id())
+        input_view.settings().set(INPUT_VIEW_TRACER_ARGS, self.TRACER_ARGS)
+
+        contents = input_view.substr(sublime.Region(0, input_view.size()))
+        update_view(output_view, *trace_code(contents, self.TRACER_ARGS))
+
+
+class StartCommand(BaseStartCommand):
+
+    pass
+
+
+class StartCanvasCommand( BaseStartCommand ):
+
+    TRACER_ARGS = ['--canvas']
+
+
+class OutputViewReplaceCommand(sublime_plugin.TextCommand):
+
+    def run(self, edit, text=''):
+        self.view.replace(edit, sublime.Region(0, self.view.size()), text)
+
+
+class InputViewEventListener(sublime_plugin.ViewEventListener):
+
+    """
+    Will execute live coding update once the timeout expires. Based on the 
+    IdleWatcher example here: http://www.sublimetext.com/docs/plugin-examples
+    """
+
+    pending = 0
+
+    @classmethod
+    def is_applicable(cls, settings):
+        return settings.has(OUTPUT_VIEW_ID)
+
+    def on_timeout_async(self):
+        self.pending = self.pending - 1
```

Comment:
Good call. Will make the change.

Reply:
Just looking at this again. I think the original example (http://www.sublimetext.com/docs/plugin-examples) is set up this way because you can't veto a callback once the timeout has been set. This means that the only way to know which is the most recent callback is to use a counter, some uuid, or look at milliseconds against the epoch or similar. If it's a bool there's no way to know how many keystrokes are still queued.

Reply:
Thanks for the explanation. Looks good to me.